codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java

```diff
@@ -0,0 +1,266 @@
+
+    private final Http2Connection connection;
+    private final Http2Connection.PropertyKey stateKey;
+    private final WriteVisitor writeVisitor = new WriteVisitor();
+    private final Deque<State> queue = new ArrayDeque<State>();
+    private final EnqueueStrategy defaultEnqueueStrategy = new EnqueueStrategy();
+
+    /**
+     * The first time a state is enqueued, it will be added to the head of the queue. After
+     * that it will be added to the tail.
+     */
+    private final EnqueueStrategy firstEnqueueStrategy = new EnqueueStrategy() {
+        @Override
+        void enqueue(State state) {
+            queue.addFirst(state);
+            state.enqueued = true;
+
+            // Switch to the default strategy so that it will be added to the tail from
+            // now on.
+            state.enqueueStrategy = defaultEnqueueStrategy;
+        }
+    };
+
+    /**
+     * The minimum number of bytes that we will attempt to allocate to a stream. This is to
+     * help improve goodput on a per-stream basis.
+     */
+    private final int minAllocationChunk;
+    private long totalStreamableBytes;
+
+    public UniformStreamByteDistributor(Http2Connection connection) {
+        this(connection, DEFAULT_MIN_ALLOCATION_CHUNK);
+    }
+
+    public UniformStreamByteDistributor(Http2Connection connection, int minAllocationChunk) {
+        this.connection = checkNotNull(connection, "connection");
+        if (minAllocationChunk <= 0) {
+            throw new IllegalArgumentException("minAllocationChunk must be > 0");
+        }
+        this.minAllocationChunk = minAllocationChunk;
+
+        // Add a state for the connection.
+        stateKey = connection.newKey();
+        Http2Stream connectionStream = connection.connectionStream();
+        connectionStream.setProperty(stateKey, new State());
+
+        // Register for notification of new streams.
+        connection.addListener(new Http2ConnectionAdapter() {
+            @Override
+            public void onStreamAdded(Http2Stream stream) {
+                stream.setProperty(stateKey, new State());
+            }
+
+            @Override
+            public void onStreamClosed(Http2Stream stream) {
+                state(stream).close();
+            }
+        });
+    }
+
+    @Override
+    public void updateStreamableBytes(StreamState streamState) {
+        State state = state(streamState.stream());
+        state.updateStreamableBytes(streamState.streamableBytes(), streamState.hasFrame());
+    }
+
+    @Override
+    public boolean distribute(int maxBytes, Writer writer) throws Http2Exception {
+        checkNotNull(writer, "writer");
+
+        // Allocate the bytes to the streams.
+        long allocated = allocateBytes(maxBytes);
+
+        long unallocatedStreamableBytes = totalStreamableBytes - allocated;
+
+        // Need to write even if maxBytes == 0 in order to handle the case of empty frames.
+        writeVisitor.writeAllocatedBytes(writer);
```

Comment:
So part of the point of this algorithm is that it is greedy and doesn't need this second pass. Do we get much benefit from not writing immediately? I would expect writing out immediately to produce simpler code (less book keeping), be more fair, and be faster (since it doesn't have to loop through all streams).

I would expect the main argument to be to allow write combining for the same stream when it is allocated bytes multiple times. However, I could easily be in favor of _not_ write combining, because write combining is unfair from a queuing standpoint. Any writes that occur here go into another queue to be written. Since we are putting an upper bound on the outbound queue this unfairness is not too big of a deal. But the point is that write combining isn't all that important, so we can take it or leave it.

Note that if you don't have something like `minAllocationChunk`, then write combining is important in order to prevent writing one byte at a time. But since this does have `minAllocationChunk`, `minAllocationChunk` is already protecting against "unreasonably small" writes.